src/config/middleware/experimental-metrics-csv-middleware.ts

```diff
@@ -0,0 +1,164 @@
+import {Middleware, MiddlewareRequestHandler} from './middleware';
+import {Metadata, StatusObject} from '@grpc/grpc-js';
+import {Message} from 'google-protobuf';
+import * as fs from 'fs';
+import {MomentoLogger, MomentoLoggerFactory} from '../logging/momento-logger';
+
+function headerFields(): Array<string> {
+  return [
+    'numActiveRequests',
+    'requestType',
+    'status',
+    'startTime',
+    'endTime',
+    'duration',
+    'requestSize',
+    'responseSize',
+  ];
+}
+
+interface RequestMetrics {
+  numActiveRequests: number;
+  requestType: string;
+  status: number;
+  startTime: number;
+  endTime: number;
+  duration: number;
+  requestSize: number;
+  responseSize: number;
+}
+
+class ExperimentalMetricsCsvMiddlewareRequestHandler
+  implements MiddlewareRequestHandler
+{
+  private readonly logger: MomentoLogger;
+  private readonly csvPath: string;
+  private readonly startTime: number;
+  private requestType: string;
+  private requestSize: number;
+  private responseStatusCode: number;
+  private responseSize: number;
+
+  private receivedResponseBody: boolean;
+  private receivedResponseStatus: boolean;
+
+  constructor(logger: MomentoLogger, csvPath: string) {
+    this.logger = logger;
+    this.csvPath = csvPath;
+    ExperimentalMetricsCsvMiddleware.numActiveRequests++;
+    this.startTime = new Date().getTime();
```

Comment:
Would you consider using the nodejs performance monitoring toolkit for measuring latency (and maybe more) instead of re-rolling it? https://nodejs.org/api/perf_hooks.html#performancenow

Reply:
the reason is because I want to see the actual epoch timestamps; there is a particular customer that I'm trying to debug and I want to be able to process the data from the csv to see how many requests are overlapping, how bursty the traffic is, etc.  I could certainly use their timer to compute the duration (and maybe I will in a follow-up before we make this "non-Experimental"), but since I am already having to compute the start/end timestamps anyway, it didn't seem worth doing it both ways?

Reply:
yeah, seems to me like probably the best plan here for measuring performance characteristics would be to integrate with the ecosystem's affordances for measuring performance. https://nodejs.org/api/perf_hooks.html#class-performanceobserver It's pretty good stuff, I was impressed reading through some of the stuff they offer in here just now. Seems difficult to replicate some of the relevant stuff in there, for example, [http2 details](https://nodejs.org/api/perf_hooks.html#http2-http2-details) which would be _awesome_ and richly informative if they are preserved by the underlying grpc library. If you want to make this csv thing too, 🤷 sure okay. It's labeled "experimental" like you say.

Reply:
thanks for bringing this to my attention.  we can definitely play with that for a future iteration.  this version is just to try to get us started on debugging an active customer issue and see what things are helpful and which aren't, but yeah i agree that's worth looking into for the next go-around.

Reply:
Works for me!